File: elixir/complex_numbers.ex
defmodule ComplexNumbers do
  @typedoc """
  In this module, complex numbers are represented as a tuple-pair containing the real and
  imaginary parts.
  For example, the real number `1` is `{1, 0}`, the imaginary number `i` is `{0, 1}` and
  the complex number `4+3i` is `{4, 3}'.
  """
  @type complex :: {float, float}

  @doc """
  Return the real part of a complex number
  """
  @spec real(a :: complex) :: float
  def real({r, _}), do: r

  @doc """
  Return the imaginary part of a complex number
  """
  @spec imaginary(a :: complex) :: float
  def imaginary({_, i}), do: i

  @doc """
  Multiply two complex numbers, or a real and a complex number
  """
  @spec mul(a :: complex | float, b :: complex | float) :: complex
  def mul({a, b}, {c, d}) do
    {a * c - b * d, b * c + a * d}
  end

  def mul({a, b}, c), do: {a * c, b * c}
  def mul(a, {b, c}), do: {a * b, a * c}

  @doc """
  Add two complex numbers, or a real and a complex number
  """
  @spec add(a :: complex | float, b :: complex | float) :: complex
  def add({a, b}, {c, d}), do: {a + c, b + d}
  def add(a, {c, d}), do: {a + c, d}
  def add({a, b}, c), do: {a + c, b}

  @doc """
  Subtract two complex numbers, or a real and a complex number
  """
  @spec sub(a :: complex | float, b :: complex | float) :: complex
  def sub({a, b}, {c, d}), do: {a - c, b - d}
  def sub(a, {c, d}), do: {a - c, -d}
  def sub({a, b}, c), do: {a - c, b}

  @doc """
  Divide two complex numbers, or a real and a complex number
  """
  @spec div(a :: complex | float, b :: complex | float) :: complex
  def div({a, b}, {c, d}) do
    {(a * c + b * d) / (c ** 2 + d ** 2), (b * c - a * d) / (c ** 2 + d ** 2)}
  end

  def div({a, b}, c), do: {a / c, b / c}

  def div(a, {b, c}) do
    d = b ** 2 + c ** 2
    {a * b / d, -a * c / d}
  end

  @doc """
  Absolute value of a complex number
  """
  @spec abs(a :: complex) :: float
  def abs({a, b}) do
    (a ** 2 + b ** 2) ** 0.5
  end

  @doc """
  Conjugate of a complex number
  """
  @spec conjugate(a :: complex) :: complex
  def conjugate({a, b}), do: {a, -b}

  @doc """
  Exponential of a complex number
  """
  @spec exp(a :: complex) :: complex
  def exp({a, b}) do
    ea = 2.718281828459045 ** a
    {ea * :math.cos(b), ea * :math.sin(b)}
  end
end

File: go/complex_numbers.go
package complexnumbers

import "math"

type Number struct {
	real float64
	imag float64
}

func (n Number) Real() float64 {
	return n.real
}

func (n Number) Imaginary() float64 {
	return n.imag
}

func (n1 Number) Add(n2 Number) Number {
	return Number{
		n1.real + n2.real,
		n1.imag + n2.imag,
	}
}

func (n1 Number) Subtract(n2 Number) Number {
	return Number{
		n1.real - n2.real,
		n1.imag - n2.imag,
	}
}

func (n1 Number) Multiply(n2 Number) Number {
	return Number{
		n1.real*n2.real - n1.imag*n2.imag,
		n1.imag*n2.real + n1.real*n2.imag,
	}
}

func (n Number) Times(factor float64) Number {
	return Number{
		n.real * factor,
		n.imag * factor,
	}
}

func (n1 Number) Divide(n2 Number) Number {
	d := n2.real*n2.real + n2.imag*n2.imag
	return Number{
		(n1.real*n2.real + n1.imag*n2.imag) / d,
		(n1.imag*n2.real - n1.real*n2.imag) / d,
	}
}

func (n Number) Conjugate() Number {
	return Number{
		n.real,
		-n.imag,
	}
}

func (n Number) Abs() float64 {
	return math.Sqrt(n.real*n.real + n.imag*n.imag)
}

func (n Number) Exp() Number {
	ea := math.Exp(n.real)
	return Number{
		ea * math.Cos(n.imag),
		ea * math.Sin(n.imag),
	}
}
